libs/utils/trace.py

```diff
@@ -886,6 +886,69 @@ def _compute_steps_duration(self, df):
 
         return df
 
+    def _windowify_steps_df(self, df, (start, end), len_column='len'):
+        """
+        Slice a dataframe to [start:end], and work on the time data so that it
+        makes sense within the interval.
+
+        Examples to make it clearer:
+
+        df is:
+        Time len state
+        15    1   1
+        16    1   0
+        17    1   1
+        18    1   0
+        -------------
+
+        _windowify_steps_df(df, (16.5, 17.5)) =>
+
+        Time len state
+        16.5  .5   0
+        17    .5   1
+
+        _windowify_steps_df(df, (16.2, 16.8)) =>
+
+        Time len state
+        16.2  .6   0
+
+        :returns: a new df that fits the above description
+        """
+        if df.empty:
+            return df
+
+        res = df[start:end].copy()
+
+        # There's a 99% probability that events won't fit snuggly within within
+        # the interval. Find the one preceding the interval, and deduce the rest
+        # mathemagically.
+        left = df[:start].tail(1)
+        right = res.tail(1)
+
```

Comment:
We can easily return here, since we need to generate just one row DF:

```python
if res.empty:
    new_delta = df[start:].head(1).index - start
    res = pd.DataFrame(df[:start].tail(1).values,
                       columns=df.columns, index=[start])
    res[col_name] = new_delta
    return res
```

Moreover, I've come up with this other implementation for the whole function:

```python
def slice(df, start, end, col_name):
    # Sanity check boundaries
    start = max(start, df.index[0])
    end = min(end, df.index[-1])
    
    # Let's start by building the first row
    values = df[:start].tail(1).values
    delta_first = df[start:].head(1).index - start
    begin_df = pd.DataFrame(values, columns=df.columns, index=[start])
    begin_df[col_name] = delta_first

    # Get a "middle" slice from the original data frame
    middle_df = df[start:end]

    # If the middle slice is empty, then we just need to first row previously computed
    if middle_df.empty:
        return begin_df

    # Otherwise, the middle slice can be appended to the first row
    final_df = begin_df.append(middle_df)

    # And finally fix the delta for the last row
    delta = df[:end].tail(1)[col_name].values[0]
    delta_last = delta + (end - df[end:].head(1).index).values[0]
    final_df.iloc[-1, final_df.columns.get_loc(col_name)] = delta_last

    return final_df.reset_index().drop_duplicates(keep='last', subset='index').set_index('index') 
```
I've tested with some corner cases... but it certainly deserves some more third party testing.
To me, it seems more easy to read... what do you think.

Reply:
Actually, your next patch test can give it a first check... did not run it although. ;-)